src/zeit/content/volume/browser/form.py

```diff
@@ -87,23 +83,38 @@ def setUpWidgets(self, *args, **kw):
             self.widgets['volume'].setRenderedValue(settings.default_volume)
 
     def add(self, object):
-        path = self.volume_location(object)
-        # The last part of the path is the filename for the volume object.
-        volume_filename = path[-1]
-        container = zeit.cms.content.add.find_or_create_folder(*path[:-1])
-        if self._check_duplicate_volume(container, volume_filename):
+        folder, filename = self._create_folder(object, object.product.location)
+        if folder is None:
             return
-        container[volume_filename] = object
-        self._created_object = container[volume_filename]
-        self._finished_add = True
+        folder[filename] = object
+        self._created_object = folder[filename]
+
+        cp_template = zeit.cms.interfaces.ICMSContent(
+            object.product.cp_template, None)
+        if zeit.content.text.interfaces.IPythonScript.providedBy(cp_template):
+            folder, filename = self._create_folder(
+                object, object.product.centerpage)
+            if folder is None:
+                return
+            folder[filename] = cp_template(volume=object)
 
-    def volume_location(self, object):
-        location = object.fill_template(object.product.location)
-        location = location.replace(zeit.cms.interfaces.ID_NAMESPACE, '')
-        return [x for x in location.split('/') if x]
+        self._finished_add = True
 
-    def _check_duplicate_volume(self, location, name):
-        if location.get(name) is not None:
+    def _create_folder(self, volume, text):
+        path, filename = self._make_path(volume, text)
+        folder = zeit.cms.content.add.find_or_create_folder(*path)
+        if self._check_duplicate_item(folder, filename):
+            return (None, None)
+        return folder, filename
+
+    def _make_path(self, volume, text):
+        uniqueId = volume.fill_template(text)
```

Comment:
Hat zwar mit diesem PR nichts zu tun, aber ich finde den Namen `fill_template` sehr ungünstig. Insbesondere jetzt wo wir noch `cp_template` haben. Ich finde `fill_template` klingt viel zu generisch, wobei die Anwendung ja sehr konkret ist. Zum Beispiel `format_location_string` ist sprechender. Statt einer Unterstrich-Methode `_fill_template` die dann als API nach außen verwendet wird (!) sollte es lieber eine simple Funktion in `zeit/content/volume/volume.py` sein, also `def` und so.